app/views/tasks/show.html.slim

```diff
@@ -573,10 +573,21 @@
             i.fa-regular.fa-clone
             =< t('common.button.duplicate')
 
-      - if policy(@task).generate_test?
-        = link_to generate_test_task_path(@task), method: :post, class: 'btn btn-important' do
-          i.fa-solid.fa-wand-magic-sparkles
-          =< t('.button.generate_test')
+      - if policy(@task).update?
+        - if policy(@task).generate_test?
```

Comment:
@kkoehn please discuss together because it is exhausting to go back and forth 😅

This way the pseudo code that you provided is implemented and it handles all the cases where a user can or cannot create a unit test :)

Reply:
to reply to this comment: https://github.com/openHPI/codeharbor/pull/1496#discussion_r1672982221

yes it is required to have them both. because while a user might input a valid API key in their profile, it could be that this API key becomes invalid at some point in the future. and so having the tests done in both places insures that this is handled

Reply:
I think we can remove `update` check from the `generate_test` policy in that case if i understood your comment correctly. Since it is checked separately. would that work?

Reply:
Rather than removing the `update` check from `generate_test?`, I would reorder the permission check here.

Isn't it similar to:

```rb
- if policy(@task).generate_test? # A user can update the task and has a working OpenAI API key stored in their profile.
- elsif policy(@task).update? # No OpenAI API key configured is configured for the user yet, but updating is allowed.
```

This would just reduce the nesting by one level.

Reply:
But a user can still not be able to update and have an openAI API key? wouldn't that not cover this case?

I am getting confused about shuffling all of this multiple times. I don't see the reason behind both of your decisions 😅 could you please elaborate more why this is better than what is already done for my understanding? :)

Reply:
One other question: There is nesting in other parts of the show file as well. so why is it better there and not in this example?

Reply:
> One other question: There is nesting in other parts of the show file as well. so why is it better there and not in this example?

Yes, there is, and that's not ideal either. The more nesting we have, the harder it is to read. It is just [a question of style](https://rubystyle.guide/#no-nested-conditionals):

> Avoid use of nested conditionals for flow of control.

In the current case, we cannot use any early return, but we can still reduce the nesting. [Here](https://stackoverflow.com/a/33817929) is a quick example and while it is not specific to Ruby, it provides the idea I am trying to apply here. Please correct me if I am wrong!

Then, regarding the other places: We try to live the [boy scouting rule](https://www.oreilly.com/library/view/97-things-every/9780596809515/ch08.html), aiming to improve code snippets along the way. Hence: If you found other places where nesting is not ideal, please feel free to improve those as well (ideally in at least a dedicated commit or -- even better -- in a dedicated PR). I do so, too, and you can expect some boy scouting improvements by me tomorrow (in dedicated PRs).

<hr>

Otherwise, back to the topic: Which two decisions are you referring to? The `update?` check within `generate_test?` and the suggestion about the less-nested conditional?

Reply:
@MrSerth Thank you for your explanations. I understand your point. And I quickly tested your logic here and it works as intended. I will implement it and push it in a couple of moments 😊
 
> ```ruby
> - if policy(@task).generate_test? # A user can update the task and has a working OpenAI API key stored in their profile.
> - elsif policy(@task).update? # No OpenAI API key configured is configured for the user yet, but updating is allowed.
> ```